### PR TITLE
V4 record set data

### PR DESCRIPTION
**Pull request by AlexBHarley**

### CODE REVIEW

File: src/presence/presence-handler.ts [hunk 1]
```diff
@@ -310,4 +312,11 @@ export class PresenceHandler {
       })
     }
   }
+
+  private onConnectionLost (): void {
```
Comment (AlexBHarley) [inline src/presence/presence-handler.ts]: here you also need to call `queryEmitter.off()` and `queryAllEmitter.off()` right?

File: src/record/record-core.ts [hunk 2]
```diff
@@ -487,21 +487,22 @@ export class RecordCore extends Emitter {
...
 
-
     if (message.action === RECORD_ACTION.VERSION_EXISTS) {
       // what kind of message is version exists?
       // this.recoverRecord(message)
       return
     }
 
     if (message.action === RECORD_ACTION.MESSAGE_DENIED) {
+      console.log(message)
```
Comment (yasserf) [inline src/record/record-core.ts]: console.log

File: src/record/record-handler.ts [hunk 3]
```diff
@@ -237,12 +241,19 @@ export class RecordHandler {
...
   public setData (recordName: string, path: string, data: any, callback: WriteAckCallback): void
-  public setData (recordName: string, args: utils.RecordSetArguments): void
   public setData (recordName: string, pathOrData: string | object, dataOrCallback: any | WriteAckCallback, callback?: WriteAckCallback): void
   public setData (recordName: string): void {
-    const { path, data, callback } = utils.normalizeSetArguments(arguments, 1)
+    const args = utils.normalizeSetArguments(arguments, 1)
+    this.sendSetData(recordName, args)
+  }
 
+  private sendSetData (recordName: string, args: utils.RecordSetArguments): void {
```
Comment (yasserf) [inline src/record/record-handler.ts]: sendSetData is a bit odd of a name.

File: src/record/record-handler.ts [hunk 4]
```diff
@@ -237,12 +241,19 @@ export class RecordHandler {
...
+    const { path, data, callback } = args
```
Comment (yasserf) [inline src/record/record-handler.ts]: destructuring can be done in header

File: src/record/write-ack-notifier.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
+import { RECORD_ACTIONS, Message } from '../../binary-protocol/src/message-constants'
+
+import { Services, Client, EVENT } from '../client'
+import { Options } from '../client-options'
+import { WriteAckCallback } from './record-core'
+
+/**
+ * Provides a scaffold for subscriptionless requests to deepstream, such as the SNAPSHOT
```
Comment (yasserf) [inline src/record/write-ack-notifier.ts]: documentation is wrong

File: src/record/write-ack-notifier.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+   * @param {String} name An identifier for the request, e.g. a record name
+   * @param {Object} callback An object with property `callback` or `resolve` and `reject`
+   *
+   * @public
+   * @returns {void}
+   */
+  public send (message: Message, callback: WriteAckCallback): void {
+    if (this.services.connection.isConnected === false) {
+      this.services.timerRegistry.requestIdleCallback(callback.bind(this, EVENT.CLIENT_OFFLINE))
+      return
```
Comment (yasserf) [inline src/record/write-ack-notifier.ts]: no elses after returns

**Pull request merged**